feat: Generics support for Imports

guybedford opened the pull request:
This adds generics support for imported JS functions in wasm-bindgen.

This PR explcitly then adds typing to `Promise<T>` and `Array<T>`, operating on `JsValue<T>`, in a way that is backwards compatible on the default `JsValue` generic.

* The macro parser is updated to support generics on imported wasm bindgen functions and types, supporting generics on function parameters and return types. The parser carefully validates that lifetimes are not used, other generic features like const are not used so we just have very limited simple generics with bounds, and these cases are tested in the macro tests. We also validate that method instrinsics require `&self<T...>` for all non-optional generics.
* The validated generics structure is stored using `syn::Generics` on the function and imported type, and then retained in the final generated function.
* To support extern "C" which does not support generics we make all generic types concrete on the ABI boundary by using JsValue itself as the concrete "top type" for all generics. That is generics are constrained to be commutative with JsValue through the introduction of JsValue casting.
* Code generation then applies these casts, carefully merging and applying generic bounds, and carefully separating `impl` bounds in the cases of static and non-static methods.

## JsValue Casts

The main casting cases to deal with for generating concrete ABI types from generics are in an imported function include the following. In all of these cases we need to obtain either `JsValue` or `&JsValue`:

* `T`, `Foo<T>` value parameters into `JsValue` via `Into<JsValue>` for simplicity
* `T`, `Foo<T>` value return into `JsValue` cast into the exact type via the new `wbg_cast`

Types that cannot be cast into JsValue cannot be supported in generics. Slices cannot be made generic, but `Array` can.

## Value types

For value typing we lean into the new typed `JsValue<T>` provided in https://github.com/wasm-bindgen/wasm-bindgen/pull/4746, which allow providing compile-time type hints that are runtime-validated through the generic type system.

### wbg_cast

Replacing the original `JsValueCast`, `wbg_cast` is used to cast return values from `JsValue` into the exact type, while following the same semantics of the static bindgen, thanks to the work here by @RReverser. The `From` type is always JsValue, while the `To` type is the type being cast into.

## Generic Types

With the codegen and casts implemented, we are then able to add generics support for the following here. More APIs can be covered in future, but I didn't want to go too far down this road here.

Some functions have generic types that can't be inferred, and in these cases they are left untyped. When a typed variant of the same function is needed I started a couple of cases like `Array::new()` and `Promise::new()` where `Array::new_t::<T>()` can be used with a `_t` suffix to access the typed variant. It seems like this could be a relatively sensible convention to me, but further feedback is welcome.

### Promise

All promise methods are now typed, with tests in `typed_promise.rs` values carry through full typing information from a `Promise.resolve()` through to its `future` and through to its then.

A new `then_map` method is added which supports a return value for then also being typed as two separate generic parameters for the promise then.

Because of the default `JsValue` type, these changes remain backwards compatible.

### Closures

Getting promises to work required solving closure generics where all transitive generic traits as required are now implemented for closures.

### Futures

When resolving promises into futures, futures are now all typed by default, and this is backwards compatible by having `JsValue` as the default generic parameter.

`future_to_promise` was not possible to implement typed without that being a breaking change, so instead a new `future_to_promise_typed` is introduced with the generic parameter.

### Array

Array and Array methods are typed where possible. Unfortunately `FnMut` is not possible to type without implementing that as a cast into an externref which changes the lifetime semantics, so generic function callbacks are not yet supported. I'd like to rather implement this as a follow-on.

### Option

When supporting generics on `Option<T>` we need to be able to obtain an `Option<T>` from a JsValue. This is done by treating null or undefined as the None case, and now implemented as a new feature.

When casting `Some<T>` there is of course intersection with `null` and `undefined`, where it might make sense to have a marker for non null JS Values further, I did explore this briefly but didn't want to go too far down this road.

At the end of the day, many functions will want generics on `pub fn foo(&Option<T>) -> Option<T>`, so supporting the generic conversion through JsValue casting for option is key to this.

### Type Errors

By using `JsValue<T>` type error handling can be deferred to the end user in whether they choose to unwrap via `.unwrap()` or `.try_unwrap()` to obtain the Rust type from the JS JsValue type.

RReverser commented on crates/backend/src/codegen.rs:
```diff
@@ -984,34 +982,74 @@ impl ToTokens for ast::ImportType {
+        let phantom_init;
+        let default_as_ref;
+
+        let where_clause = generics::where_clause(&self.generics);
```
These custom utils seem to duplicate what [`syn::Generics::split_for_impl`](https://docs.rs/syn/latest/syn/struct.Generics.html#method.split_for_impl) is doing.

You might want to use it instead for simplicity.

guybedford replied:
Thanks, will look into this.

guybedford replied:
I've added using `split_for_impl` here instead.

guybedford commented:
I've rebased this branch to the latest and refactored `JsValueCast` to instead use the new `wbg_cast`.

All seems to mostly work, except there is a new bug with stack handling which I'll investigate further in another week or so. If anyone wants to take a look at the bug in the meantime please feel free.

RReverser commented on src/cast.rs:
```diff
@@ -154,3 +155,36 @@ where
+/// live for as long as the function call only.
+///
+/// This trait is automatically implemented for any type supporting `AsRef<JsValue>`.
+pub trait JsUpcastRef<T = JsValue> {
```
I remember you initially added this machinery for references when you thought we can't accept both references and plain values in the same signature.

Now that we figured it out, can we leave this trait and slots implementation out of this PR? It feels like a large chunk of complexity here comes just from this reference upcasting, even though I imagine in real world scenarios it's going to be relatively uncommon for someone to want to pass `&i32` instead of `i32` and so on, and removing it from this PR would reduce scope to just generics for types we support today.

And if it's not uncommon, we can always add it separately, as being able to pass references to more types is not strictly tied to generics support.

guybedford replied:
While that is possible, the problem is the requirement for type specialization on the generic - where a singular specialization is possible on `&T` as `$JsValue` requiring no further type hints making it backwards compatible.

guybedford replied:
In case that wasn't clear - if `T` can be either `&JsValue` or `JsValue` then more type information is required from the usage context, and thus this requirement on the type is a breaking change for existing usage.

RReverser replied:
I don't understand, can you show an example that would work today but break with such generic?

guybedford replied:
Actually I think the problem is simpler:

```
extern "C" { 
    #[wasm_bindgen(static_method_of = Promise)]
    pub fn resolve(obj: &JsValue) -> Promise;
}
```

Can only be represented by the generic:

```
extern "C" { 
    #[wasm_bindgen(static_method_of = Promise)]
    pub fn resolve<T>(obj: &T) -> Promise<T>;
}
```

It cannot be represented by the generic:

```
extern "C" { 
    #[wasm_bindgen(static_method_of = Promise)]
    pub fn resolve<T>(obj: T) -> Promise<T>;
}
```

as when we generate the specialization by substituting JsValue that would be a breaking change.

RReverser replied:
Ah wait I understand what you mean. Hmm. Need to think about this.

I'd really like to avoid having to use references for primitives. Maybe there is some solution with custom trait that would have T::Output that would unwrap reference automatically...

Let's discuss more next week.

guybedford replied:
This doesn't create a requirement for using references for primitives - most APIs should use value types for primitives, and also support `&T` or `T` interchangeably. But we need to special case `&T` for this backwards compatibility.

Ensuring `T` supports a ref case could be something we could add for new APIs that want to support own or ref, but that would require a separate design. If that's the case you're interested in happy to discuss it further.

guybedford replied:
`JsUpcastRef` has now been removed and replaced with `&JsValue<T>`.
